### ISSUE TITLE
Change Request: Flat Config does not support FileEnumerator

### ISSUE THREAD

**Issue by ljharb** [eslint/eslint#18087]
### ESLint version

v8

### What problem do you want to solve?

eslint-plugin-import uses FileEnumerator in its `no-unused-modules` rule, for the purpose of gathering a list of files are not eslintignored, or ignored by the user's rule config, for checking if any exports or modules are unused.

It seems that in flat config, this capability does not exist.

### What do you think is the correct solution?

Something that may work nicely is a new method on `context` provided to rules, that can achieve the same goal, but I have no idea if this makes sense for eslint or not.

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

This is the sole remaining blocker (afaik) to eslint-plugin-import supporting Flat Config, after which I plan to do a breaking change to drop older eslint versions, which is something quite a lot of users have been asking for.

**Comment by nzakas** [comment]
> eslint-plugin-import uses FileEnumerator in its `no-unused-modules` rule, for the purpose of gathering a list of files are not eslintignored, or ignored by the user's rule config, for checking if any exports or modules are unused.

If I'm reading this correctly, your use case is basically to answer the question "is this file ignored by ESLint?" So perhaps a method like `isFileIgnored()` where you can pass in the file path would address your needs?

**Comment by ljharb** [comment]
Sort of - that would require us to enumerate every file on disk and ask eslint about each one, because we also don't know where the eslint "root" is. We specifically need a list of lintable files.

**Comment by nzakas** [comment]
Okay, then I feel like there's not enough information in the original description to really understand how you're using `FileEnumerator`. So, can you add more detail?

And when you say "root", what is it you're referring to? (Note that `context.cwd` is available already.)

**Comment by ljharb** [comment]
By `root`, i mean `root: true` in eslintrc (i'm not sure if that concept still exists in flat config - maybe the cwd is always the root now?)

What we need is a list or iterator of lintable files (so we can crawl them all, parse their imports and exports, and figure out which are unused).

**Comment by nzakas** [comment]
Okay, I'm going to try asking this again: can you please explain *exactly* the approach/algorithm that you're using right now to do this? We need to know:

1. How are you currently determining `root`?
2. Is that where you're searching from or somewhere else?
3. Are you only parsing the files you find referenced in an `import` statement, or all of the lintable files?
4. Are files parsed all up front or on-demand (or something else)?

There isn't going to be an exact replacement for `FileEnumerator`, so we need to understand the problem you're trying to solve, plus the way you already solved it, in order to figure out the approach going forward.

**Comment by ljharb** [comment]
oh sure, sorry if i misunderstood :-)

The code that uses FileEnumerator is here: https://github.com/import-js/eslint-plugin-import/blob/7a21f7e10f18c04473faadca94928af6b8e28009/src/rules/no-unused-modules.js#L19-L64 and https://github.com/import-js/eslint-plugin-import/blob/7a21f7e10f18c04473faadca94928af6b8e28009/src/rules/no-unused-modules.js#L173-L187

The rule's config takes a list of globs, and a list of exclusion patterns, which we apply to narrow down the files provided by the Enumerator.

The rule requires the globs all be relative to the cwd, so any root determination beyond that is up to eslint via the Enumerator.

We're parsing all lintable files; the user defines which extensions in their plugin settings in eslint config.

Files are all parsed "on demand" in the sense that it's when the rule is run, but "up front" in that it's the first thing the rule does (if we don't cache it for all future in-process rule runs after that, then oops, we should).

**Comment by nzakas** [comment]
That's helpful, thanks.

So am I understanding you correctly in that if I do something like `npx eslint foo.js`, `no-unused-modules` will still glob and parse all of the files specified in the config? Can you provide an example config and walk me through that?

Looking through the code, it seems like perhaps you could replace `FileEnumerator` with a generic glob utility, and then something like a `context.isFileIgnored()` method would allow the same behavior you have now. It looks like you're already defaulting to searching `process.cwd()` if `src` isn't specified, which you can get from `context.cwd`.

It doesn't appear that you need the location of the ESLint config file for any reason (what we called `root` in eslintrc), or am I missing something?

**Comment by ljharb** [comment]
yes, that's right - unfortunately i don't have an example config available (they're in previous employers' codebases).

using a generic glob utility would be "fine" but there'd be no guarantee it matches the glob semantics eslint uses, now and in the future.

(when i said "root" i meant the `root` boolean property in eslintrc; i certainly would prefer not to ever interact with config files directly)

**Comment by sam3k** [comment]
Per 2024-02-08 `tsc-meeting`, @nzakas @mdjermanovic will follow up on this issue.

**Comment by nzakas** [comment]
> using a generic glob utility would be "fine" but there'd be no guarantee it matches the glob semantics eslint uses, now and in the future.

I think you can get pretty close. Internally, we have `isDirectoryIgnored()` and `isFileIgnored()` that we use to filter glob results, which you can see here:
https://github.com/eslint/eslint/blob/15c143f96ef164943fd3d39b5ad79d9a4a40de8f/lib/eslint/eslint-helpers.js#L285-L333

The rest of the logic around filtering there is solely for the benefit of being able to output useful errors on the CLI, which is clearly something you wouldn't need to do.

So, I think if we exposed `isDirectoryIgnored()` and `isFileIgnored()` to rules, you should be able to mimic what we're doing here.

**Comment by ljharb** [comment]
I agree, that would work, thank you. The only remaining downside is that I'd basically have to copy-paste that logic, and if eslint ever changed it, it'd be out of sync. Is there no way to get direct access to eslint's glob results in the first place, even if unfiltered?

**Comment by nzakas** [comment]
> The only remaining downside is that I'd basically have to copy-paste that logic, and if eslint ever changed it, it'd be out of sync.

That is a risk, for sure, but given that we use the same `isDirectoryIgnored()` and `isFileIgnored()`, those are the most likely things that would change and you'd get those passed through to you. 

> Is there no way to get direct access to eslint's glob results in the first place, even if unfiltered?

Unfiltered would be disastrous because you'd end up getting everything in every `node_modules` (that's part of the filtering).

The glob results of the current run probably won't work for you because that won't give you all of the files in the project. For example, if someone runs `npx eslint foo.js`, then the result of the search will only be `/path/to/foo.js`. We only search based on the pattern that was passed in, and if I'm understanding you correctly, you'd need more than that if you want to check if a module is unused anywhere in the project.

As I said, this isn't really functionality we want to encourage in rules, so we're trying to make the surface area to get this working for you as small as possible.

(At some point, I'd like to look at a higher level at what it might look like for ESLint to understand the dependency graph of a project.)

**Comment by ljharb** [comment]
True, basically what we'd want is "the list of paths to be linted as if someone ran `eslint .` in the cwd" (which is always the way i configure every project; i personally consider passing a path to be an antipattern).

**Comment by nzakas** [comment]
`--ext` doesn't exist in flat config -- all that data is in the config file and therefore is respected in `isDirectoryIgnored()` and `isFileIgnored()`. 

Let me put together a quick POC that you can try out.

**Comment by ljharb** [comment]
Ah, thanks, that simplifies it a bit. Look forward to seeing what you come up with :-)

**Comment by nzakas** [comment]
Here's the branch:
https://github.com/eslint/eslint/tree/issue18087-poc

The way I have this set up is as a `context.session` object that has `isDirectoryIgnored()` and `isFileIgnored()` as methods. (This is only available when running in flat config mode.) Each method expects an absolute path to be passed in.

Let me know how that works.

**Comment by ljharb** [comment]
So, if someone does `eslint foo.js`, will anything that's *not* foo.js be considered ignored?

**Comment by ljharb** [comment]
ok - so is there any way for a rule to know that those options were provided?

(It feels more and more like this is something eslint should be abstracting away, rather than a plugin having to reconstruct eslint's logic)

**Comment by nzakas** [comment]
* `--config` just changes the config file being used and that [becomes the config array](https://github.com/eslint/eslint/blob/f95cd27679eef228173e27e170429c9710c939b3/lib/eslint/eslint.js#L331) 
* `--ignore-pattern` is [merged into the config array](https://github.com/eslint/eslint/blob/f95cd27679eef228173e27e170429c9710c939b3/lib/eslint/eslint.js#L406)
* `--no-ignore` is also used to [calculate the config array](https://github.com/eslint/eslint/blob/f95cd27679eef228173e27e170429c9710c939b3/lib/eslint/eslint.js#L390).
 
So `isDirectoryIgnored()` and `isFileIgnored()` takes all of this into account.

**Comment by ljharb** [comment]
ok awesome!

**Comment by ljharb** [comment]
no, not just yet - I'll try to make some time today!

**Comment by nzakas** [comment]
ping @ljharb -- we really need your feedback on this approach to help with release planning

**Comment by ljharb** [comment]
I'm trying to look into it, but since I don't use flat config anywhere yet, it's not something I can quickly test :-/

However, I feel pretty confident the approach will work, it's just unfortunate that the globbing logic has to be replicated.

**Comment by ljharb** [comment]
Mainly because almost all of my repos are stuck on eslint v8.8.0 due to https://github.com/eslint/eslint/pull/15595, and I no longer have access to a corporate repo, so I'm not actually using the no-unused-modules rule. I'm not really sure how I can get someone to try it out prior to it being published, unfortunately.

**Comment by JoshuaKGoldberg** [comment]
I had some time to look into this. Mainly good news and a bit of bad news.

Good news: I was able to get a working `no-unused-modules` rule based on a limited copy & paste of `FileEnumerator` that receives `isDirectoryIgnored` and `isFileIgnored` instead of config logic. So I think that confirms the hope in this thread: that should be enough to support flat config (at the cost of some copy & paste). Reference PR here: https://github.com/import-js/eslint-plugin-import/pull/2967.

Bad news: the specific `isFileIgnored` implementation in -> https://github.com/eslint/eslint/tree/issue18087-poc https://github.com/eslint/eslint/commit/4c4c1f4b973b3f2039813025d047a98801c95c59 seems to be giving inconsistent results for whether a file is ignored. So there's some work to be done there. I didn't have time to dig deeply yet - but @nzakas I'm guessing you have much more context than me anyway. I put a write up in https://github.com/JoshuaKGoldberg/repros/blob/0458e256d49b3887c2b5e7da27ee932d1a91ffa7/README.md.

**Comment by nzakas** [comment]
@ljharb Unfortunately, we're not going to be able to ship anything until you've verified that this approach works. The ball is in your court now and it's up to you as to whether or not this makes it in for the final v9.0.0.

@JoshuaKGoldberg My hunch is that it's in the implementation, because we use `isFileIgnored()` frequently in the core and have a ton of tests validating its behavior. I'd suggest instead of trying to implement a new `FileEnumerator`, that you just use `fswalk` like we do: https://github.com/eslint/eslint/blob/15c143f96ef164943fd3d39b5ad79d9a4a40de8f/lib/eslint/eslint-helpers.js#L19. It does have a [walkSync](https://www.npmjs.com/package/@nodelib/fs.walk#walksyncpath-optionsorsettings) method that should work for this purpose.

**Comment by alexrecuenco** [comment]
This might help as an example, @ljharb . I was updating the template I have for typescript-based projects

In the commit `20bfe75` I had set-up "import" in a flat config, and I was getting some errors from it, maybe some config like this might help you explore the issues?

Let me know if I can help simplify this any further. 

https://github.com/alexrecuenco/typescript-eslint-prettier-jest-example/pull/15/commits/20bfe7579ed88eb1030883a4f48dad97158b6019

**Comment by controversial** [comment]
https://github.com/import-js/eslint-plugin-import/pull/3018 was merged today including an implementation that can consume these proposed APIs!

**Comment by nzakas** [comment]
@controversial unfortunately, it looks like we're not going to be able to move forward with those new APIs due to some other changes in the core. We'll need to think further about this.

**Comment by ljharb** [comment]
@nzakas what are those changes?

**Comment by nzakas** [comment]
Primarily this: https://github.com/eslint/rfcs/pull/120

But I think overall we need to rethink what it is you're doing in `eslint-plugin-import`. It's one of the most common causes of slow ESLint runs, so I don't think trying to make a drop-in replacement is actually the best idea.

**Comment by ljharb** [comment]
I’m happy to solve it with an alternative implementation, but preventing cycles is hugely important, and I’m not sure how else to do it besides checking every file (altho a built-in “pre-lint stage” where the plugin builds up a dep graph for files would likely be much more performant)

**Comment by nzakas** [comment]
Unfortunately, ESLint isn't (currently) built for the use case of checking every file in a project every time it's run. Many times, people are running ESLint on just a single file (in their IDE) or a few files (in a precommit check). I'd really like to avoid checking every file in those use cases.

What we may be talking about here is a different mode that ESLint runs in and a different type of rule that runs across multiple files instead of just a single one. We are also talking about a lot of work to get there.

Alternatively, we could just recommend a dedicated tool like `knip` for this purpose.

**Comment by ljharb** [comment]
Multiple use cases have been put forth for a "pre lint" stage, where rules that are about to run get to register callbacks or similar - in our case, that would allow rules that need the dep graph to populate it, once, and cache it across relevant rules; the graph would not be constructed when it's not needed; and `TIMING` values would be able to accurately report rule times vs pre-lint-stage times. Is that not a viable path forward? eslint core still wouldn't have to officially support running on multiple files.

**Comment by guillaumebrunerie** [comment]
> Alternatively, we could just recommend a dedicated tool like `knip` for this purpose.

Knip is a good replacement for the `import/no-unused-modules` rule (except for the fact that it doesn't have an LSP server), but it does not check import cycles. 

For checking cycles, yet another tool like Madge would be needed.

**Comment by ljharb** [comment]
Either way, having to use another tool for a statically knowable thing is a really huge and painful ask - imo eslint should be handling all static analysis, just like it has been doing.

**Comment by nzakas** [comment]
> Multiple use cases have been put forth for a "pre lint" stage, where rules that are about to run get to register callbacks or similar - in our case, that would allow rules that need the dep graph to populate it, once, and cache it across relevant rules; the graph would not be constructed when it's not needed; and `TIMING` values would be able to accurately report rule times vs pre-lint-stage times. Is that not a viable path forward? eslint core still wouldn't have to officially support running on multiple files.

It would be helpful for you to open up a separate issue to discuss this further so it doesn't get lost in these comments. It's a bit difficult to understand what it is you're envisioning from just a couple of sentences.

FWIW, I did open an RFC for prelint plugins:
https://github.com/eslint/rfcs/pull/105

I haven't had time to revisit it, and I'm not sure it's the same as what you're thinking, which is why it would be good to pull this line out into a separate issue.

> Either way, having to use another tool for a statically knowable thing is a really huge and painful ask - imo eslint should be handling all static analysis, just like it has been doing.

I don't think ESLint needs to solve every static analysis problem, especially ones that it wasn't designed to handle.

**Comment by timmywil** [comment]
> I don't think ESLint needs to solve every static analysis problem, especially ones that it wasn't designed to handle.

Fair enough, but the problem with that in this case is that eslint was already being used to handle something and now it can't be, and therefore has broken a very popular plugin with no migration plan. It's the kind of breaking change that should have been avoided before v9 was released, and yes, even if FileEnumerator was an internal, undocumented API.
